src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/MetadataBuilder.Heaps.cs

```diff
@@ -212,17 +222,27 @@ public int GetHeapOffset(UserStringHandle handle)
             return MetadataTokens.GetHeapOffset(handle);
         }
 
-        public UserStringHandle GetUserString(string str)
+        /// <exception cref="ImageFormatLimitationException">The remaining space on the heap is too small to fit the string.</exception>
+        public UserStringHandle GetOrAddUserString(string str)
         {
             int index;
             if (!_userStrings.TryGetValue(str, out index))
             {
                 Debug.Assert(!_streamsAreComplete);
 
-                index = _userStringWriter.Position + _userStringHeapStartOffset;
-                _userStrings.Add(str, index);
-                _userStringWriter.WriteCompressedInteger(str.Length * 2 + 1);
+                int startPosition = _userStringWriter.Position;
+                int encodedLength = str.Length * 2 + 1;
+                index = startPosition + _userStringHeapStartOffset;
 
+                // Native metadata emitter allows strings to exceed the heap size limit as long 
+                // as the index is within the limits (see https://github.com/dotnet/roslyn/issues/9852)
```

Comment:
While the GitHub issue linked explains in more detail, I find this comment a bit hard to understand. I'd make it clearer about what this implementation does. As it is, it's hard to tell if this is trying to say that we should be compatible with native reader or deliberately strengthen the limit.